### ISSUE TITLE
React deprecated replaceState for ES6 classes

### ISSUE THREAD

**Issue by michael** [substance/archivist-writer#14]
WE use replaceState a lot to modify the app state in a clean way, using setState we would have a lot of garbage around in the state object. What should we do here?

**Comment by michael** [comment]
actually it's not deprecated in ES6 classes it's not available, so we can not use it at all.

**Comment by michael** [comment]
I did a bit of research and it looks like that the motivation is that they want to encourage the best practice that state has an initial signature and whenever it changes it stays within that signature.

See: https://github.com/facebook/react/issues/3236

Probably in our case we could solve the problem by only using setState and modelling our writer state more explicitly.

Instead of: 

``` js
{
  contextId: "entities",
  entityId: "entity_12"
}
```

We could do: 

``` js
{
  contextId: "entities",
  params: {entityId: "entity_12"}
}
```

Bwah i'm not sure i like it... because we also have the modal scope:

``` js
{
  contextId: "entities",
  params: {entityId: "entity_12"},
  modal: {
    contextId: "manageEntities",
    params: {"entityType": "location"}
  }
}
```

I'm not really sure what is the advantage because you can anyway nest your state deeply.

**Comment by michael** [comment]
Probably we should just store everything in one property appState, and leave the structure untouched.. then we can have replaceState behaviour.

``` js
this.setState({appState: {
  contextId: "entities",
  entityId: "entity_12"
});
```

I really don't get their reasoning behind it, because nothing is won. I would even say replace state should be the default.. (when calling setState) while a separate updateState merges the state.

**Comment by obuchtala** [comment]
I agree the default behavior should be replaceState.

Just to summarize: to actually have an explicit replace we need an extra wrapper:

```
setState({ state: {} })
```

This would always be like a replace state, right?
But with the incredibly ugly extra layer...

**Comment by obuchtala** [comment]
or we write a helper that takes the current state keys, with 'undefined' values, and extending it with what we want to set.

```
setState(_.extend({a: undefined, b:undefined},{a: "foo", c: "bar"})
```

**Comment by michael** [comment]
yep i guess that will be our solution for the time being. I did not dive into alternatives such as using Immutable.js, where i think react is aware of it's immutability and replaces the state with the new one. However Immutable.js is also a few thousand lines of code long.. and not sure it's overkill for our use-case. We'd need to evaluate it's positive impact on the easy implementation of componentShouldUpdate, where we don't need to do a deep check anymore.. just compare the oldState and newState. Maybe it's worth it.

**Comment by obuchtala** [comment]
And how would that work:

```
setState(Object.freeze({...}))
```

?

**Comment by michael** [comment]
no idea.. but that would be smart if that worked!

**Comment by michael** [comment]
hey did you see this fellow? #9 (just in case you missed the email)
